libmobilecoin/src/chacha20_rng.rs

```diff
@@ -0,0 +1,188 @@
+///
+/// # Arguments
+///
+/// * `long_val` - an unsigned 64 bit value to use as the rng seed
+///
+/// # Errors
+///
+/// * `LibMcError::Poison`
+#[no_mangle]
+pub extern "C" fn mc_chacha20_rng_create_with_long(
+    long_val: u64,
+    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
+) -> FfiOptOwnedPtr<Mutex<McChaCha20Rng>> {
+    ffi_boundary_with_error(out_error, || {
+        Ok(Mutex::new(McChaCha20Rng::seed_from_u64(long_val)))
+    })
+}
+
+/// Returns a new ChaCha20Rng instance initialized with the
+/// seed value provided by the bytes data, which must be at
+/// least 32 bytes (only the first 32 bytes will be used)
+///
+/// # Arguments
+///
+/// * `bytes` - 32 bytes of data to use as the rng seed
+///
+/// # Errors
+///
+/// * `LibMcError::Poison`
+#[no_mangle]
+pub extern "C" fn mc_chacha20_rng_create_with_bytes(
+    bytes: FfiRefPtr<McBuffer>,
+    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
+) -> FfiOptOwnedPtr<Mutex<McChaCha20Rng>> {
+    ffi_boundary_with_error(out_error, || {
+        let bytes: [u8; 32] = bytes
+            .as_slice()
+            .try_into()
+            .expect("seed size must be 32 bytes");
+        Ok(Mutex::new(McChaCha20Rng::from_seed(bytes)))
+    })
+}
+
+/// Returns the current word_pos of the ChaCha20Rng instance
+///
+/// # Arguments
+///
+/// * `chacha20_rng` - must be a valid ChaCha20Rng
+/// * `out_word_pos` - pointer to buffer of 128 bytes where the current
+///   chacha20_rng wordpos will be returned
+///
+/// # Errors
+///
+/// * `LibMcError::Poison`
+#[no_mangle]
+pub extern "C" fn mc_chacha20_rng_get_word_pos(
+    chacha20_rng: FfiMutPtr<Mutex<McChaCha20Rng>>,
+    out_word_pos: FfiMutPtr<McMutableBuffer>,
+    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
+) -> bool {
+    ffi_boundary_with_error(out_error, || {
+        let word_pos = chacha20_rng.lock()?.get_word_pos();
+        let mc_u128 = McU128::from_u128(word_pos);
+
+        let out_word_pos = out_word_pos.into_mut().as_slice_mut_of_len(16)?;
+
+        out_word_pos.copy_from_slice(&mc_u128.bytes);
+
+        Ok(())
+    })
+}
+
+/// Sets the current word_pos of the ChaCha20Rng instance
+///
+/// /// # Arguments
+///
+/// * `chacha20_rng` - must be a valid ChaCha20Rng
+/// * `out_word_pos` - pointer to buffer of 128 bytes where the current
+///   chacha20_rng wordpos will be returned
+///
+/// # Errors
+///
+/// * `LibMcError::Poison`
+#[no_mangle]
+pub extern "C" fn mc_chacha20_rng_set_word_pos(
+    chacha20_rng: FfiMutPtr<Mutex<McChaCha20Rng>>,
+    bytes: FfiRefPtr<McBuffer>,
+    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
+) -> bool {
+    ffi_boundary_with_error(out_error, || {
+        let mc_u128 = McU128 {
+            bytes: bytes
+                .as_slice()
+                .try_into()
+                .expect("word_pos length is not exaclty 16 bytes"),
+        };
+        let word_pos = mc_u128.to_u128();
+
+        chacha20_rng.lock()?.set_word_pos(word_pos);
+
```

Comment:
This may not compile/pass lint as-is, but you don't actually need the intermediate `McU128` type here, you can just use the u128 type directly, e.g.:

```suggestion
        chacha20_rng
            .lock()?
            .set_word_pos(u128::from_be_bytes(bytes
                .as_slice()
                .try_into()
                .expect("word_pos length is not exaclty 16 bytes"),
            );
```

Reply:
Updating